File: src/targets/host/wire_deploy_directory.py
"""Checks for the wire-server-deploy directory on the admin host.

In Wire-managed clusters, the admin host has ~/wire-server-deploy/ with
ansible playbooks, helm values, and deployment artifacts. If the cluster
is Wire-managed but this directory is missing, something's wrong.
"""

from __future__ import annotations

# External
import json
from typing import Any

# Ours
from src.lib.base_target import BaseTarget, NotApplicableError


class WireDeployDirectory(BaseTarget):
    """Check for wire-server-deploy directory on the admin host.

    Only runs when wire_managed_cluster is true.
    """

    # Needs SSH access to the admin host
    requires_ssh: bool = True

    @property
    def description(self) -> str:
        """What this target checks."""
        return "Wire-server-deploy directory on admin host"

    @property
    def explanation(self) -> str:
        """Why this matters."""
        return (
            "Wire-managed clusters use the wire-server-deploy repository on the admin "
            "host for deployment. It contains ansible playbooks, helm values, inventory "
            "files, and deployment scripts."
        )

    @property
    def unit(self) -> str:
        """No unit — returns structured JSON string."""
        return ""

    def collect(self) -> str:
        """Check for wire-server-deploy directory and list contents.

        Returns:
            JSON string with directory existence and contents.

        Raises:
            NotApplicableError: If cluster is not Wire-managed.
        """
        if not self.config.options.wire_managed_cluster:
            raise NotApplicableError("Cluster is not Wire-managed (no ansible layer expected)")

        self.terminal.step("Checking for wire-server-deploy directory on admin host...")

        # Check if the directory exists and list top-level contents
        result_str, _parsed = self.run_command_on_host(
            self.config.admin_host.ip,
            "ls -1 ~/wire-server-deploy/ 2>/dev/null || echo '__NOT_FOUND__'"
        )

        output: str = result_str.stdout.strip() if result_str else ""

        if "__NOT_FOUND__" in output or not output:
            result: dict[str, Any] = {
                "exists": False,
                "subdirectories": [],
                "inventory_dirs": [],
            }
            self._health_info = "wire-server-deploy directory NOT found on admin host"
            return json.dumps(result)

        # Parse directory listing
        subdirectories: list[str] = [line.strip() for line in output.split("\n") if line.strip()]

        # Check for inventory directories
        inv_result_str, _parsed2 = self.run_command_on_host(
            self.config.admin_host.ip,
            "ls -1 ~/wire-server-deploy/ansible/inventory/ 2>/dev/null || echo ''"
        )
        inv_output: str = inv_result_str.stdout.strip() if inv_result_str else ""
        inventory_dirs: list[str] = [d.strip() for d in inv_output.split("\n") if d.strip()]

        result = {
            "exists": True,
            "subdirectories": subdirectories,
            "inventory_dirs": inventory_dirs,
        }

        self._health_info = (
            f"wire-server-deploy found: {len(subdirectories)} items, "
            f"inventory: {', '.join(inventory_dirs) if inventory_dirs else 'none'}"
        )

        return json.dumps(result)

File: ui/src/checkers/operations/wire_deploy_directory.ts
/**
 * Checks wire-server-deploy directory on the admin host.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class WireDeployDirectoryChecker extends BaseChecker {
    readonly path: string = 'operations/wire_deploy_directory'
    readonly name: string = 'Wire-server-deploy directory'
    readonly category: string = 'Operations / Tooling'
    readonly interest = 'Setup' as const
    readonly requires_ssh: boolean = true
    readonly explanation: string = 'Wire-managed clusters have `~/wire-server-deploy/` on the admin host with ansible playbooks, helm values, and deployment artifacts.'

    check(data: DataLookup): CheckResult {
        if (data.config && !data.config.options.wire_managed_cluster) {
            return { status: 'not_applicable', status_reason: 'Not a Wire-managed cluster.' }
        }

        const point = data.get('host/wire_deploy_directory')
        if (!point?.value) return { status: 'gather_failure', status_reason: 'Wire deploy directory check not collected.' }

        let parsed: Record<string, unknown> | null = null
        try { parsed = JSON.parse(String(point.value)) } catch { /* ignore */ }
        if (!parsed) return { status: 'gather_failure', status_reason: 'Could not parse data.' }

        if (!(parsed.exists as boolean)) {
            return { status: 'warning', status_reason: 'Wire-managed cluster declared but `~/wire-server-deploy/` **not found** on admin host.', raw_output: point.raw_output }
        }

        const subdirs: string[] = (parsed.subdirectories as string[]) ?? []
        const inv_dirs: string[] = (parsed.inventory_dirs as string[]) ?? []

        return {
            status: 'healthy',
            status_reason: `Wire-server-deploy directory found: ${subdirs.length} items${inv_dirs.length > 0 ? `, inventory: ${inv_dirs.join(', ')}` : ''}.`,
            display_value: `found (${subdirs.length} items)`,
            raw_output: point.raw_output,
        }
    }
}

export default WireDeployDirectoryChecker
